source/RoslynAnalyzers/MessageContractAnalyzers.cs

```diff
@@ -0,0 +1,222 @@
+                        CheckProperties(context, typeDec);
+                        MessageTypes_MustHaveXmlDocComments(context, typeDec);
+                    }
+                }
+                
+                // request/command/response specific
+                switch (requestOrCommandDec?.Identifier.Text)
+                {
+                    case TypeNameICommand:
+                        if (CommandTypes_MustBeNamedCorrectly(context, typeDec))
+                            CommandTypes_MustHaveCorrectlyNamedResponseTypes(context, typeDec, requestOrCommandDec); // the expected response name depends on the command name; so only run this check if the CommandName was good
+                        break;
+
+                    case TypeNameIRequest:
+                        if (RequestTypes_MustBeNamedCorrectly(context, typeDec))
+                            RequestTypes_MustHaveCorrectlyNamedResponseTypes(context, typeDec, requestOrCommandDec); // the expected response name depends on the request name; so only run this check if the RequestName was good
+                        break;
+                }
+
+                if (responseEventOrResourceDec?.Identifier.Text == TypeNameIEvent)
+                {
+                    EventTypes_MustBeNamedCorrectly(context, typeDec);
+                }
+            }
+        }
+        
+        static bool CheckProperties(SyntaxNodeAnalysisContext context, TypeDeclarationSyntax typeDec)
+        {
+            var result = true;
+            foreach (var propDec in typeDec.DescendantNodes().OfType<PropertyDeclarationSyntax>())
+            {
+                // only validate public properties
+                if (!propDec.Modifiers.Any(m => m.IsKind(SyntaxKind.PublicKeyword))) continue;
+
+                // we have a number of places where we treat collection types differently; front-load that info
+                var isCollectionType = IsCollectionType(context, propDec);
+                var required = GetRequiredState(propDec);
+
+                // if anything returns false, propagate false outwards
+                result &= PropertiesOnMessageTypes_MustBeMutable(context, propDec);
+                result &= RequiredPropertiesOnMessageTypes_MustNotBeNullable(context, propDec, required);
+                result &= OptionalPropertiesOnMessageTypes_ExceptForCollections_MustBeNullable(context, propDec, required, isCollectionType);
+                result &= MessageTypes_MustInstantiateCollections(context, propDec, required, isCollectionType);
+                result &= PropertiesOnMessageTypes_MustHaveAtLeastOneValidationAttribute(context, propDec, required);
+                result &= SpaceIdPropertiesOnMessageTypes_MustBeOfTypeSpaceId(context, propDec);
+                result &= IdPropertiesOnMessageTypes_MustBeACaseInsensitiveStringTinyType(context, propDec);
+            }
+
+            return result;
+        }
+        
+        // ----- helpers --------------- 
```

Comment:
I get nervous about having helpers which are intended to be private and only work within the narrow niche of one class, being pulled out into other classes and files because a class is big.

Extracting the helpers advertises that they're fit for re-use in other places, which they are not 😞